flow/layers/layer.h

```diff
@@ -119,47 +92,54 @@ struct PrerollContext {
-  // The internal_nodes_canvas is a SkNWayCanvas which is used by non leaf
-  // and applies the operations to all canvases.
-  // The leaf_nodes_canvas is the "current" canvas and is used by leaf
-  // layers.
-  SkCanvas* internal_nodes_canvas;
-  SkCanvas* leaf_nodes_canvas;
+  // a platform view on iOS) during the paint traversal we apply any state
+  // changes which affect children (i.e. saveLayer attributes) to the
+  // state_stack and any local rendering state changes for leaf layers to
+  // the canvas or builder.
+  // When we switch a canvas or builder (when painting a PlatformViewLayer)
+  // the new canvas receives all of the stateful changes from the state_stack
+  // to put it into the exact same state that the outgoing canvas had at the
+  // time it was swapped out.
+  // The state stack lazily applies saveLayer calls to its current canvas,
+  // allowing leaf layers to report that they can handle rendering some of
+  // its state attributes themselves via the |applyState| method.
+  LayerStateStack& state_stack;
+  SkCanvas* canvas;
+  DisplayListBuilder* builder = nullptr;
+
   GrDirectContext* gr_context;
   SkColorSpace* dst_color_space;
   ExternalViewEmbedder* view_embedder;
   const Stopwatch& raster_time;
   const Stopwatch& ui_time;
   std::shared_ptr<TextureRegistry> texture_registry;
   const RasterCache* raster_cache;
-  const bool checkerboard_offscreen_layers;
   const float frame_device_pixel_ratio = 1.0f;
 
   // Snapshot store to collect leaf layer snapshots. The store is non-null
   // only when leaf layer tracing is enabled.
   LayerSnapshotStore* layer_snapshot_store = nullptr;
   bool enable_leaf_layer_tracing = false;
-
-  // The following value should be used to modulate the opacity of the
-  // layer during |Paint|. If the layer does not set the corresponding
-  // |layer_can_inherit_opacity()| flag, then this value should always
-  // be |SK_Scalar1|. The value is to be applied as if by using a
-  // |saveLayer| with an |SkPaint| initialized to this alphaf value and
-  // a |kSrcOver| blend mode.
-  SkScalar inherited_opacity = SK_Scalar1;
-  DisplayListBuilder* leaf_nodes_builder = nullptr;
-  DisplayListBuilderMultiplexer* builder_multiplexer = nullptr;
 };
 
 // Represents a single composited layer. Created on the UI thread but then
 // subquently used on the Rasterizer thread.
 class Layer {
  public:
+  // The state attribute flags that represent which attributes a
+  // layer can render if it plans to use a saveLayer call in its
+  // |Paint| method.
+  static constexpr int SAVE_LAYER_RENDER_FLAGS =
+      LayerStateStack::CALLER_CAN_APPLY_OPACITY |
+      LayerStateStack::CALLER_CAN_APPLY_COLOR_FILTER |
+      LayerStateStack::CALLER_CAN_APPLY_IMAGE_FILTER;
+
+  // The state attribute flags that represent which attributes a
+  // layer can render if it will be rendering its content/children
+  // from a cached representation.
+  static constexpr int RASTER_CACHE_RENDER_FLAGS =
```

Comment:
Could we also fold color filter or image filter into drawing from cache?

Reply:
Layer caching was somewhat disrupted by this new mechanism. I'll have to resurrect it first.

Reply:
Layer caching seems to be working OK, even though I thought I had left some wires hanging out of it...

Reply:
My goal here is to achieve mostly parity with existing layer caching and leave enhancing the caching to new combinations for a future PR...